components/tooltip/tooltip.js

```diff
@@ -128,6 +308,57 @@ class Tooltip extends RtlMixin(LitElement) {
 		this._target.addEventListener('mouseleave', this._onTargetMouseLeave);
 		this._target.addEventListener('focus', this._onTargetFocus);
 		this._target.addEventListener('blur', this._onTargetBlur);
+
+		this._targetSizeObserver = new ResizeObserver(this._onTargetResize);
+		this._targetSizeObserver.observe(this._target);
+	}
+
+	_computeAvailableSpaces(targetRect, spaceAround) {
+		const verticalWidth = Math.max(spaceAround.left + targetRect.width + spaceAround.right, 0);
+		const horizontalHeight = Math.max(spaceAround.above + targetRect.height + spaceAround.below, 0);
+		const spaces = [
+			{ dir: 'top', width: verticalWidth, height: Math.max(spaceAround.above - this.offset, 0) },
+			{ dir: 'bottom', width: verticalWidth, height: Math.max(spaceAround.below - this.offset, 0) },
+			{ dir: 'right', width: Math.max(spaceAround.right - this.offset, 0), height: horizontalHeight },
+			{ dir: 'left', width: Math.max(spaceAround.left - this.offset, 0), height: horizontalHeight }
+		];
+		if (this.getAttribute('dir') === 'rtl') {
+			const tmp = spaces[2];
+			spaces[2] = spaces[3];
+			spaces[3] = tmp;
+		}
+		return spaces;
+	}
+
+	_computeTooltipShift(centerDelta, spaceLeft, spaceRight) {
+
```

Comment:
A few of these methods seem like they could move out of the class itself and just live up above in the file. That way they stay hidden to consumers and we also likely get a small performance boost as the JS classes that get created over and over again are smaller.